examples/step-50/step-50.cc

```diff
@@ -14,912 +14,1427 @@
+                       const unsigned int /*component*/ = 0) const override;
 
-    MatrixType system_matrix;
+  template <typename number>
+  VectorizedArray<number> value(const Point<dim, VectorizedArray<number>> &p,
+                                const unsigned int /*component*/ = 0) const;
 
-    IndexSet locally_relevant_set;
+  template <typename number>
+  number average_value(const std::vector<Point<dim, number>> &points) const;
+};
 
-    AffineConstraints<double> constraints;
 
-    VectorType solution;
-    VectorType system_rhs;
+template <int dim>
+double Coefficient<dim>::value(const Point<dim> &p, const unsigned int) const
+{
+  for (int d = 0; d < dim; ++d)
+    {
+      if (p[d] < -0.5)
+        return 100.0;
+    }
+  return 1.0;
+}
 
-    const unsigned int degree;
 
-    // Finally we are storing the various parallel multigrid matrices. Our
-    // problem is self-adjoint, so the interface matrices are the transpose
-    // of each other, so we only need to compute/store them once.
-    MGLevelObject<MatrixType> mg_matrices;
-    MGLevelObject<MatrixType> mg_interface_matrices;
-    //
-    MGConstrainedDoFs mg_constrained_dofs;
-  };
+template <int dim>
+template <typename number>
+VectorizedArray<number>
+Coefficient<dim>::value(const Point<dim, VectorizedArray<number>> &p,
+                        const unsigned int) const
+{
+  VectorizedArray<number> return_value = VectorizedArray<number>();
+  for (unsigned int i = 0; i < VectorizedArray<number>::size(); ++i)
+    {
+      bool found = false;
+      for (int d = 0; d < dim; ++d)
+        if (p[d][i] < -0.5)
+          {
+            return_value[i] = 100.0;
+            found           = true;
+            break;
+          }
+
+      if (!found)
+        return_value[i] = 1.0;
+    }
+
+  return return_value;
+}
 
 
+template <int dim>
+template <typename number>
+number Coefficient<dim>::average_value(
+  const std::vector<Point<dim, number>> &points) const
+{
+  number average(0);
+  for (unsigned int i = 0; i < points.size(); ++i)
+    average += value(points[i]);
+  average /= points.size();
 
-  // @sect3{Nonconstant coefficients}
+  return average;
+}
 
-  // The implementation of nonconstant
-  // coefficients is copied verbatim
-  // from step-5 and step-6:
 
-  template <int dim>
-  class Coefficient : public Function<dim>
-  {
-  public:
-    virtual double value(const Point<dim> & p,
-                         const unsigned int component = 0) const override;
 
-    virtual void value_list(const std::vector<Point<dim>> &points,
-                            std::vector<double> &          values,
-                            const unsigned int component = 0) const override;
-  };
+/**
+ * Matrix-free Laplace operator
+ */
+template <int dim, int fe_degree, typename number>
+class LaplaceOperator
+  : public MatrixFreeOperators::Base<dim,
+                                     LinearAlgebra::distributed::Vector<number>>
+{
```

Comment:
Can't you use `MatrixFreeOperators::LaplaceOperator` from [here](https://dealii.org/developer/doxygen/deal.II/classMatrixFreeOperators_1_1LaplaceOperator.html)? Since this tutorial is merely using matrix-free infrastructure, I would imagine that you do not really need to show more than necessary. In that class, you have everything and only need to fill a `std::shared_ptr<Table<2,VectorizedArrayType>>` with the coefficients. Here you use the average, but that could be fit into that table in any case (at a slight loss of efficiency because the same data is replicated at all quadrature points).

Reply:
Let me know what you think about my approach. I added a `Coefficient::create_coefficient_table()` which fills the `std::shared_ptr<Table<2,VectorizedArrayType>>`.

The reason we were using the self-defined operator was so that we wouldn't have to store duplicate viscosity on quadrature, but this should hopefully not be too expensive to store (certainly not for this tutorial).

Reply:
> Perhaps a cellwise coefficient could be added to the LaplaceOperator class? Not sure how useful that would be.

I agree that such a coefficient would make sense. Let me open a PR.

Reply:
A cellwise coefficient makes a lot of sense - I have opened the pull request #10025 for that.

Reply:
Perfect. I'm now using in in `create_coefficient_table()`